plutus-playground-client/src/AjaxUtils.purs

```diff
@@ -1,20 +1,89 @@
 module AjaxUtils where
 
 import Bootstrap (alertDanger_)
+import Control.Alt ((<|>))
+import Control.Monad.Except.Trans (ExceptT(ExceptT), runExceptT)
 import Control.Monad.Reader.Class (class MonadAsk, ask)
-import Control.MonadPlus ((=<<))
+import Control.MonadPlus (empty, (=<<))
 import Data.Argonaut.Core (Json)
 import Data.Argonaut.Generic.Aeson (decodeJson)
+import Data.Argonaut.Generic.Aeson as Aeson
+import Data.Argonaut.Generic.Decode (genericDecodeJson)
+import Data.Argonaut.Generic.Options (Options(..))
 import Data.Argonaut.Parser (jsonParser)
 import Data.Array (intercalate)
-import Data.Either (Either(..))
-import Data.Generic (class Generic)
+import Data.Either (Either(Right, Left), note)
+import Data.Generic (class Generic, GenericSignature(SigProd), GenericSpine(SProd), fromSpine, isValidSpine, toSignature, toSpine)
+import Data.List as List
+import Data.List.NonEmpty as NonEmpty
+import Data.Maybe (Maybe(..), fromMaybe)
+import Gist (GistId)
 import Halogen.HTML (HTML, br_, div_, pre_, text)
 import Language.Haskell.Interpreter (CompilationError(..))
 import Network.HTTP.StatusCode (StatusCode(..))
-import Prelude (bind, pure, show, ($), (<$>), (<>), (>>>))
+import Playground.API (TokenId)
+import Playground.Server (SPParams_(..))
+import Prelude (bind, pure, show, unit, ($), (<$>), (<>), (>>>))
 import Servant.PureScript.Affjax (AjaxError, ErrorDescription(ConnectionError, DecodingError, ParsingError, UnexpectedHTTPStatus), runAjaxError)
-import Servant.PureScript.Settings (SPSettingsDecodeJson_(..), SPSettingsEncodeJson_(..), SPSettings_(..))
+import Servant.PureScript.Settings (SPSettingsDecodeJson_(..), SPSettingsEncodeJson_(..), SPSettingsToUrlPiece_(..), SPSettings_(..), URLPiece, defaultSettings, gDefaultToURLPiece)
+import Type.Proxy (Proxy(..))
+
+ajaxSettings :: SPSettings_ SPParams_
+ajaxSettings = SPSettings_ $ settings
+  { toURLPiece = SPSettingsToUrlPiece_ gCustomToURLPiece
+  , decodeJson = SPSettingsDecodeJson_ (genericDecodeJson $ Options $ options {userDecoding = userDecoding})
+  }
+  where
+    SPSettings_ settings = defaultSettings $ SPParams_ { baseURL: "/api/" }
+    Options options = Aeson.options
+
+ourDecodeJson :: forall a. Generic a => Json -> Either String a
+ourDecodeJson =
+  let (SPSettings_ {decodeJson: SPSettingsDecodeJson_ decodeJson}) = ajaxSettings
+  in decodeJson
+
+userDecoding :: Options -> GenericSignature -> Json -> Maybe (Either String GenericSpine)
```

Comment:
`Maybe (Either String A)` if the `String` represents an error, why the `Maybe`?

Reply:
The `Maybe` represents a "pass". If you return `Nothing`, it means you're not going to try to decode this, and the next decoder in the chain should be asked.

Reply:
Sure but you can do this with `Either` as it has an `Alt` instance. Maybe this is needed to fit in with existing types though? So do you ever return a `Just . Left` ?

Reply:
Yeah, `Just . Left` would be legitimate, and would mean, "I took responsibility
for decoding this, and it failed."

Reply:
(Of course you can _try_ decoding, fail, and then decide to pass with a `Nothing`. Both are valid, depending on what behaviour you want. Quite a good API IMO.)

Reply:
but I mean what would the value of the `Left` be and where would it be used?

Reply:
Aren't these `Left` values just disappearing in the end though? Or I suppose you can use them to keep a track of what failed?